tests: random lifespan threads test

niewim19 opened the pull request:
Testing if multiple threads with pseudorandom lifespan start and close correctly on imxrt106x

JIRA: CI-65

<!--- Provide a general summary of your changes in the Title above -->

## Description
New test addition for imxrt106x

## Motivation and Context
To test if adding test with random lifespan crashes the imxrt106x platform

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

<!--- In case of breaking change - please advice here what needs to be done in dependent projects. -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
- [ ] Already covered by automatic testing.
- [ ] New test added: (add PR link here).
- [x] Tested by hand on: imxrt106x.

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing linter checks and tests passed.
- [ ] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [ ] This PR needs additional PRs to work (list the PRs, preferably in merge-order).
- [x] I will merge this PR by myself when appropriate.

niewim19 commented:
By `I  will merge this PR by myself when appropriate.` I mean that this PR is waiting for syspage name size correction as for now it is named incorrectly:
-is named: `tthr_rand`
-should be: test-thread-rand
But the correct name exceeds the maximum application name length in syspage or imxrt106x

nalajcie commented on threads/tthr_rand.c:
```diff
@@ -0,0 +1,83 @@
+	uint16_t id;
+} threadss_self_t;
+
+int fin;
```
static

niewim19 replied:
Used static

nalajcie reviewed (changes requested):
not sure if intentionally left as a draft, but got here via email and did review anyway :).

I've pointed some things that need to be fixed.

Other notes:
 - when You fix coding style issue reported by reviewdod - please "resolve conversation" so the reviewer would know it was taken care of
 - I believe the the binary names can be longer now (see https://github.com/phoenix-rtos/phoenix-rtos-kernel/pull/199)
 - this test should also work on ia32 (there is nothing platform-specific here) so please include it in the testing.

niewim19 commented on threads/tthr_rand.c:
```diff
@@ -0,0 +1,83 @@
+int fin;
+
+
+void test_thread(void *arg)
```
Used static

niewim19 commented on threads/tthr_rand.c:
```diff
@@ -0,0 +1,83 @@
+	/* 	all threads should end in under one second */
+	sleep(1);
+	/* printf("End main thread %d\n",fin); */
+	TEST_ASSERT_TRUE(fin == cthread);
```
You mean `TEST_ASSER_EQUAL_INT` :) Fixed, thanks!

niewim19 commented on threads/tthr_rand.c:
```diff
@@ -0,0 +1,83 @@
+		usleep(100);
+	}
+
+	/* 	all threads should end in under one second */
```
fixed, thanks!

niewim19 commented on threads/tthr_rand.c:
```diff
@@ -0,0 +1,83 @@
+#include "unity_fixture.h"
+
+
+typedef struct threadss_self {
```
- yes, this was a typo that was propagated by mistake
 - removed struct name

niewim19 commented on threads/tthr_rand.c:
```diff
@@ -0,0 +1,83 @@
+void test_thread(void *arg)
+{
+	threadss_self_t *self = (threadss_self_t *)arg;
+	/* printf("Starting thr: %d for %d ms\n", self->id, self->time); */
```
Commented code was removed

niewim19 commented on threads/tthr_rand.c:
```diff
@@ -0,0 +1,83 @@
+	}
+
+	/* 	all threads should end in under one second */
+	sleep(1);
```
`threadJoin` used

niewim19 commented on threads/tthr_rand.c:
```diff
@@ -0,0 +1,83 @@
+	/* assign threads parameters (quick pseudorandom) */
+	for (int i = 0; i < cthread; i++) {
+		params[i].id = i; /* assign thread id */
+		n = (n * seed) % 200;
```
`rand()` used. Also used srand to ensure the same seed for every run

niewim19 commented on threads/tthr_rand.c:
```diff
@@ -0,0 +1,83 @@
+{
+	int cthread = 10, seed = 1087, n = 107;
+	threadss_self_t params[cthread];
+	void *stack[cthread * 512];
```
Type changed to `char` and switched to static allocation

niewim19 commented on threads/tthr_rand.c:
```diff
@@ -0,0 +1,83 @@
+	/* printf("Starting thr: %d for %d ms\n", self->id, self->time); */
+	usleep(1000 * self->time);
+	/* printf("Ending %d\n", self->id); */
+	fin++;
```
Used __sync_fetch_and_add to reassure correct behavior

nalajcie replied:
obsolete.

niewim19 commented on threads/threads.yaml:
```diff
@@ -0,0 +1,7 @@
+          exec: tthr_rand
+          type: unit
+          targets:
+              value: [armv7m7-imxrt106x]
```
Yes. I started with a test only for imxrt106x, but it can be run also on ia32

kemonats commented on threads/test_thread_rand.c:
```diff
@@ -0,0 +1,88 @@
+static int cthread = 10, fin = 0;
+
+
+static char stack[10 * 512];
```
```suggestion
static char stack[cthread * 512];
```

niewim19 replied:
As `cthread` and `stack` are now stored in struct i am not able to define size of `stack` base on `cthread` value. Switching to dynamic allocation allows for it but I want to stay simple and do not allocate and then free memory for stack.

nalajcie replied:
obsolete now.

kemonats commented on threads/test_thread_rand.c:
```diff
@@ -0,0 +1,88 @@
+	}
+
+	/* all threads should end in under one second */
+	for (i = 0; i < 10; i++) {
```
```suggestion
	for (i = 0; i < cthread; i++) {
```

niewim19 replied:
Thanks! Changed.

pawelpisarczyk commented:
I think that this test should be located in proc subdirectory. General remark - please, before start writing a code, perform a short analysis of current project structure, coding convention etc. This will reduce our effort in making reviews.

pawelpisarczyk commented on threads/test_thread_rand.c:
```diff
@@ -0,0 +1,88 @@
+#include <string.h>
```
File should start with the properly constructed label as it is described in documentation.

niewim19 replied:
Yes, my mistake, label added.

nalajcie replied:
label needs fixing, please read what You're copying!

pawelpisarczyk commented on threads/test_thread_rand.c:
```diff
@@ -0,0 +1,88 @@
+	uint16_t time;
+	uint16_t id;
+} threads_self_t;
+
```
I don't understand the reason of using types with defined bit size. TIme and id should be defined using common C types e.g. unsigned int.

niewim19 replied:
At first i was experimenting with `uint16_t` instead of `char` while defining stack. Changed to unsigned int.

pawelpisarczyk commented on threads/test_thread_rand.c:
```diff
@@ -0,0 +1,88 @@
+static int cthread = 10, fin = 0;
+
+
+static char stack[10 * 512];
```
We agreed to use common structures holding common. global attributes attributes. For code clarity is better to use two dimensional array.

niewim19 replied:
Two dimensional array implemented. `thread_common` structure also created.

pawelpisarczyk commented on threads/test_thread_rand.c:
```diff
@@ -0,0 +1,88 @@
+{
+	threads_self_t *self = (threads_self_t *)arg;
+	usleep(1000 * self->time);
+	__sync_fetch_and_add(&fin, 1);
```
Please don't use function embedded into the GNU CC compiler. Please synchronize global counter using mutex defined in _common structure. I don't understand why we should rely on GNU CC primitives. In this particular test we are not aiming to write high-performance code minimizing context switching. Rather we would like to test  correctness of constructing and destroying threads.

niewim19 replied:
Changed global counter synchronization. Now using `mutexLock()` and `mutexUnlock()` with lock stored in `thread_common` structure.

pawelpisarczyk commented on threads/test_thread_rand.c:
```diff
@@ -0,0 +1,88 @@
+	/* run threads */
+	for (i = 0; i < cthread; i++) {
+		beginthread(test_thread, 6, &(stack[i * 512]), 512, &(params[i]));
+		usleep(100);
```
What is the reason of adding here 100us delay?

niewim19 replied:
Yes it is unnecessary here. Removed.

pawelpisarczyk commented on threads/test_thread_rand.c:
```diff
@@ -0,0 +1,88 @@
+
+TEST(threads_rand, test_1)
+{
+	threads_self_t params[cthread];
```
This is the inconsequence. Thread stacks are defined globally in BSS, but thread parameters are define locally on stack? Why?

niewim19 replied:
Stack was moved to BSS because it was necessary to even store it. Parameters were left as local. Now with common struct `params` they are also stored in BSS for consistency. 

Only variables `joined` and `i` were left on stack as they scope is limited only for use in main test function `TEST(threads_rand, test_1)`

pawelpisarczyk commented on threads/test_thread_rand.c:
```diff
@@ -0,0 +1,88 @@
+	/* assign threads parameters */
+	for (i = 0; i < cthread; i++) {
+		params[i].id = i;
+		params[i].time = 100 + rand() % 200; /* assign thread lifespan in milliseconds */
```
I think that time randomization range should be defined properly using defined scale projected on RAND_MAX macro.

niewim19 replied:
Yes it would be more professional, but i would rather stay with this simple approach with just `%` and added constant value as we only need an integer and we do not really bother with edge cases. This saves us couple of casts (from integer to double and then finally to integer for `sleep` functions) and keeps the code clean.

pawelpisarczyk commented on threads/threads.yaml:
```diff
@@ -0,0 +1,7 @@
+          exec: test_thread_rand
+          type: unit
+          targets:
+              include: [armv7m7-imxrt106x]
```
This test should be accessible on all hardware platforms. Marek mentioned this in his review.

niewim19 replied:
Previous Marek`s note was resolved with change from
```
targets:
    value: [armv7m7-imxrt106x]
```
to
```
targets:
    include: [armv7m7-imxrt106x]
```
as `value:` specifies the only architecture to run test on, and `include:` adds it to test targets.

nalajcie replied:
still not being run on ia32-generic

pawelpisarczyk reviewed (dismissed):
Please change the code according to our requests.

nalajcie commented on proc/test_thread_rand.c:
```diff
@@ -0,0 +1,108 @@
+/*
+ * Phoenix-RTOS
+ *
+ * phoenix-rtos-tests
```
Please put a short description here. What You're testing?

niewim19 replied:
Description added. However some of other test files need checking cause they lack short description.

nalajcie commented on proc/test_thread_rand.c:
```diff
@@ -0,0 +1,108 @@
+typedef struct {
+	unsigned int cthread;
+	unsigned int fin;
+	char stack[10][512];
```
aligned properly by accident, please add explicit alignment

niewim19 replied:
Added explicit alignment for `stack`

nalajcie commented on proc/test_thread_rand.c:
```diff
@@ -0,0 +1,108 @@
+
+	/* run threads */
+	for (i = 0; i < threads_common.cthread; i++)
+		beginthread(test_thread, 6, &(threads_common.stack[i]), 512, &(threads_common.params[i]));
```
maybe proper `sizeof` instead of hardcoded value?

niewim19 replied:
Used `sizeof`. Thanks

nalajcie commented on proc/test_thread_rand.c:
```diff
@@ -0,0 +1,108 @@
+
+	srand(7);
+	/* assign threads parameters */
+	for (i = 0; i < threads_common.cthread; i++) {
```
could be done in a single `for` loop with starting the thread

niewim19 replied:
This loop was divided into two for better clarity, but you are right: one loop is shorter and still clear to read.

nalajcie commented on proc/threads.yaml:
```diff
@@ -0,0 +1,7 @@
+test:
+    tests:
+        - name: test_threads_rand
```
why `threads` / `thread`?

niewim19 replied:
My mistake. Changed plural form to singular `thread`

nalajcie commented on proc/test_thread_rand.c:
```diff
@@ -0,0 +1,108 @@
+
+
+typedef struct {
+	unsigned int cthread;
```
`cthread` is misleading, when I change the value below the program will break.

I would prefer to make it `static const int` - then it can be used as a const (so also in structs definition):
```c
static const int cthread = 10;

typedef struct {
	unsigned int fin;
	char stack[cthread][512];
	[...]
}
```

You can also group (sort) struct fields according to their usage (eg. `fin` and `fin_lock`, `stack` and `params`),

nalajcie replied:
sorry, my bad, not possible with GCC and pure C (clang actually compiles that even if it's not C standard-compilant). 

We either need to use `#define` or switch to dynamic allocation.

niewim19 replied:
I moved `stack` into `thread_self_t` and separated out `time` and `id` into `thread_params_t` structure. Now we can define the amount of threads in one place and still have static allocation. 
Also the split was done to avoid passing thread `stack` as parameter to itself.  

As for `cthread` being misleading: yes its true. Now `cthread` variable is only local for main test run and not global and its value is based on size of `threads` array.

nalajcie commented on proc/test_thread_rand.c:
```diff
@@ -0,0 +1,108 @@
+	threads_self_t *self = (threads_self_t *)arg;
+	usleep(1000 * self->time);
+
+	mutexLock(threads_common.fin_lock);
```
I don't see this mutex being created anywhere. Does this even work!?

niewim19 replied:
Surprisingly it works. Added proper creation and destroy of mutex with `mutexCreate()` and `resourceDestroy()`.

nalajcie replied:
Actually - You don't know as You're not checking the result code of `mutexLock` so You don't know if it succeeded.

Please add checking result code to all native calls (`mutex*`, `begin/endthread`, `resourceDestroy`)

nalajcie reviewed (changes requested):
Some code fixes are still needed.

I don't see this test being run automatically - **this is a requirement for merging this test**:
https://github.com/phoenix-rtos/phoenix-rtos-tests/runs/2892877920?check_suite_focus=true#step:6:47

niewim19 commented on proc/test_thread_rand.c:
```diff
@@ -0,0 +1,108 @@
+ * phoenix-rtos-tests
+ *
+ * Copyright 2021 Phoenix Systems
+ * Author: Hubert Buczynski
```
Yes, my bad. Changed.

niewim19 commented:
Tests on `ia32` were blocked because of _not matching the pattern_ filename of `.yaml` script. It was changed to `test.yaml`.

niewim19 commented:
Removed unnecessary includes

nalajcie reviewed (changes requested):
Please add checking result code to all native calls (`mutex*`, `begin/endthread`, `resourceDestroy`).

I would like to see more thread-related unit tests in the same file/binary in the future (we can change file names then).

niewim19 commented:
Added checking the state. Went straight for checking the output value using _unity_ functions (not using if-s) as we have to mainly catch errors, not process them.

`endthread()` returns void so no check for it.

nalajcie commented on proc/test_thread_rand.c:
```diff
@@ -0,0 +1,121 @@
+	usleep(params->time * 1000);
+
+	err = mutexLock(threads_common.fin_lock);
+	TEST_ASSERT_EQUAL_INT(err, EOK);
```
please note that the `TEST_ASSERT_EQUAL_INT` is defined as: `TEST_ASSERT_EQUAL_INT(expected, actual)`, so You need to change the order of params for some of the calls).

niewim19 replied:
Checked and changed if necessary. Thanks.

jsarzy replied:
I think we should rather not use `TEST_ASSERT*` in threads. It uses `longjmp` and/or sets a global state which is not protected by any synchronization primitives.

niewim19 replied:
Ok, so the reasonable option is to in some way escape the error from thread. I can make another parameter in `thread_params_t` for that.

niewim19 replied:
@nalajcie , is this an issue?

nalajcie replied:
Oh, sorry, didn't know that it uses `longjmp`. As doing `longjmp` from different thread that `setjmp` is undefined behavior - it may (and probably will) result in kernel panic or some strange errors (You may test it by Yourself ny failing on purpose).

For now we can add some result code in `thread_params_t` and verify it in the main thread, maybe we would come up with better solution for multithreading support in `unity` framework in the future.

niewim19 replied:
Thanks! Going on with new field in `thread_params_t` then.

nalajcie commented on proc/test_thread_rand.c:
```diff
@@ -0,0 +1,121 @@
+
+	/* all threads should end in under one second */
+	for (i = 0; i < cthread; i++) {
+		threadJoin(0);
```
please check return code

niewim19 replied:
`threadJoin(0);` here returns id (not local) of thread. https://github.com/phoenix-rtos/phoenix-rtos-kernel/blob/964cd68eac3da0175fd98abc25cb65538dbf1825/proc/threads.c#L1138-L1146

So there is nothing to check the return value against.

niewim19 replied:
Ok. After consulting the error check is added.

nalajcie reviewed (commented):
Looks much better, pointed out small issues to be fixed.

pawelpisarczyk commented:
niewim19 what is the status of solving issues?

pawelpisarczyk commented on proc/test_thread_rand.c:
```diff
@@ -0,0 +1,123 @@
+
+typedef struct {
+	unsigned int fin;
+	handle_t fin_lock;
```
Please change this name to word without `_` to be inline with our coding standards.

niewim19 replied:
Yes, that name is not within our coding standards. I will fix that.

pawelpisarczyk commented on proc/test.yaml:
```diff
@@ -0,0 +1,7 @@
+          exec: test_thread_rand
+          type: unit
+          targets:
+              include: [armv7m7-imxrt106x]
```
Marek and me asked you to compile this test on all targets. When this issue will be addressed?

niewim19 replied:
Test is now build and tested on all targets that our runner supports (ia32 and imxrt106x). When the automatic tests will be available for imxrt106x (soon) even that include will not be necessary. The same goes with other automatic tests currently available.

The term `include:` does not specify _the only_ architecture but it adds it. Even if I specify all architectures here, tests do not build everywhere - only on imxrt106x, ia32 and host-pc

nalajcie replied:
@pawelpisarczyk this is taken care of, `targets` is a dict which supports attributes (`include`, `exclude` and `values`). The default value for `values` is `DEFAULT_TARGETS` (currently `ia32-generic` as all tests should work on this platform).

The final set of targets is `targets['values'] + targets['include'] - targets['exclude']` (yeah, seems unnaturally complex, but common pattern in YAML configs) - works great when You have multiple tests in single YAML and You globally set targets for all of them and then just include/exclude one or two for a certain target)

pawelpisarczyk commented on proc/test_thread_rand.c:
```diff
@@ -0,0 +1,123 @@
+	/* assign threads parameters */
+	for (i = 0; i < cthread; i++) {
+		threads_common.threads[i].params.id = i;
+		threads_common.threads[i].params.time = 100 + rand() % 200; /* assign thread lifespan in milliseconds */
```
I've asked you to use RAND_MAX constant to scale the randomization space to defined by constant.

niewim19 replied:
Ok, added random number generation based on RAND_MAX and rand().

Also, for better randomization:
 - switched to microseconds
 - added `timemax` and `timemin` to `threads_common`

niewim19 commented:
Test runner fail **not** due to `test_thread_rand` fail

nalajcie commented:
> Test runner fail **not** due to `test_thread_rand` fail. Also I am awaiting commentary on using `TEST_ASSERT*` inside threads. When that issue will be resolved i will push final commit together with reviewdog suggestion.

There were some changes in new toolchain and respective docker images. Please rebase Your branch on top of current master - then the tests will work again.

niewim19 commented:
Status of this PR recent changes:
 - added proper error handling from `thread*` and `mutex*` functions, which are thread safe.
 - changed the resolution of time variable by changing to microseconds and using `RAND_MAX` for greater randomization

nalajcie reviewed (approved):
You could use `float` instead of `double`, but apart from that te test is ready to merge.

Other improvements might be done as more threading unit tests will be added to this test group.
